Bug/66611 activerecord statementinvalid in get api v3 queries queriesapi queries id and get api v3 queries queriesbyprojectapi projects id queries default

ulferts opened the pull request:
# Ticket

https://community.openproject.org/wp/66611

# What are you trying to accomplish?

Enable baseline to work flawlessly with the changes to the WP.allowed_to scope.

# What approach did you choose and why?

Baseline before worked by manipulating the AR/arel structure to replace references to the model table (e.g. `work_packages`) by references to the journals (e.g. `work_package_journals`). With the changes to `WorkPackages.allowed_to`, the work packages eligible to be filtered are determined beforehand, on the current work packages, based on the current `project_id`. But when filtering historic data, the `project_id` might have been a different value. Therefore, the filtering for the allowed work packages has to happen on the data the query is run for, which is stored in the data journals (e.g `work_package_journals`).

The filtering for allowed work packages happens in a series of CTEs:

```
                ...
		WITH
			"member_projects" AS (
				(
					(
						SELECT
							"projects"."id",
							"members"."entity_id"
						FROM
							"members"
							INNER JOIN "projects" ON "projects"."active" = TRUE
							AND "members"."project_id" = "projects"."id"
							INNER JOIN "enabled_modules" ON "projects"."id" = "enabled_modules"."project_id"
							AND "enabled_modules"."name" IN ('work_package_tracking')
							AND "projects"."active" = TRUE
							INNER JOIN "member_roles" ON "member_roles"."member_id" = "members"."id"
							INNER JOIN "roles" ON "roles"."id" = "member_roles"."role_id"
							INNER JOIN "role_permissions" ON "roles"."id" = "role_permissions"."role_id"
							AND (
								FALSE
								OR "role_permissions"."permission" = 'view_work_packages'
								AND "enabled_modules"."name" = 'work_package_tracking'
							)
						WHERE
							"members"."user_id" = 86481
							AND (
								"members"."entity_type" IS NULL
								OR "members"."entity_type" = 'WorkPackage'
							)
					)
					UNION ALL
					(
						SELECT
							"projects"."id",
							NULL AS ENTITY_ID
						FROM
							"projects"
							INNER JOIN "enabled_modules" ON "projects"."id" = "enabled_modules"."project_id"
							AND "enabled_modules"."name" IN ('work_package_tracking')
							AND "projects"."active" = TRUE
							INNER JOIN "roles" ON "roles"."builtin" = 1
							AND "projects"."active"
							AND "projects"."public"
							AND NOT (
								EXISTS (
									SELECT
										1
									FROM
										"members"
									WHERE
										"members"."user_id" = 86481
										AND "members"."entity_type" IS NULL
										AND "members"."project_id" = "projects"."id"
								)
							)
							INNER JOIN "role_permissions" ON "roles"."id" = "role_permissions"."role_id"
							AND (
								FALSE
								OR "role_permissions"."permission" = 'view_work_packages'
								AND "enabled_modules"."name" = 'work_package_tracking'
							)
					)
				)
			),
			"entity_member_projects" AS (
				SELECT
					*
				FROM
					MEMBER_PROJECTS
				WHERE
					ENTITY_ID IS NOT NULL
			),
			"project_member_projects" AS (
				SELECT
					*
				FROM
					MEMBER_PROJECTS
				WHERE
					ENTITY_ID IS NULL
			),
			"entity_member_projects_without_duplicates" AS (
				SELECT
					*
				FROM
					ENTITY_MEMBER_PROJECTS
				WHERE
					NOT EXISTS (
						SELECT
							1
						FROM
							PROJECT_MEMBER_PROJECTS
						WHERE
							PROJECT_MEMBER_PROJECTS.ID = ENTITY_MEMBER_PROJECTS.ID
					)
			),
			"allowed_by_projects_and_work_packages" AS (
				SELECT
					*
				FROM
					WORK_PACKAGES
				WHERE
					PROJECT_ID IN (
						SELECT
							ID
						FROM
							MEMBER_PROJECTS
					)
					AND NOT EXISTS (
						SELECT
							1
						FROM
							ENTITY_MEMBER_PROJECTS_WITHOUT_DUPLICATES
						WHERE
							ENTITY_MEMBER_PROJECTS_WITHOUT_DUPLICATES.ID = WORK_PACKAGES.PROJECT_ID
							AND ENTITY_MEMBER_PROJECTS_WITHOUT_DUPLICATES.ENTITY_ID != WORK_PACKAGES.ID
					)
			)

...
```

Before this PR, the resulting "allowed_by_projects_and_work_packages" was then used to retrieve all work packages by using `WorkPackage.from("allowed_by_projects_and_work_packages work_packages"`. But this fails because of the manipulations done for the historic query.

By switching this to a simple `WHERE EXISTS (SELECT 1 from allowed_by_projects_and_work_packages ...)` the query itself is valid again but fails to filter on the historic project_id as described above.

Instead of attempting to work this change into the existing structure another approach is taken. What the query understands as `work_packages` is redefined by placing a CTE under that name at the very top of the CTEs. Doing so will lead to all later CTEs as well as non CTE queries within the statement to understand `work_packages` as whatever the CTE defines instead of the also existing `work_packages` table. The redefinition of `work_packages` is then a join statement on the journals of the timestamp(s) in question:

```
WITH
			"work_packages" AS (
				SELECT
					JOURNALS.JOURNABLE_ID AS ID,
					JOURNALS.CREATED_AT,
					JOURNALS.UPDATED_AT,
					WORK_PACKAGE_JOURNALS.TYPE_ID,
					WORK_PACKAGE_JOURNALS.PROJECT_ID,
					WORK_PACKAGE_JOURNALS.SUBJECT,
					WORK_PACKAGE_JOURNALS.DESCRIPTION,
					WORK_PACKAGE_JOURNALS.DUE_DATE,
					WORK_PACKAGE_JOURNALS.CATEGORY_ID,
					WORK_PACKAGE_JOURNALS.STATUS_ID,
					WORK_PACKAGE_JOURNALS.ASSIGNED_TO_ID,
					WORK_PACKAGE_JOURNALS.PRIORITY_ID,
					WORK_PACKAGE_JOURNALS.VERSION_ID,
					WORK_PACKAGE_JOURNALS.AUTHOR_ID,
					WORK_PACKAGE_JOURNALS.DONE_RATIO,
					WORK_PACKAGE_JOURNALS.ESTIMATED_HOURS,
					WORK_PACKAGE_JOURNALS.START_DATE,
					WORK_PACKAGE_JOURNALS.PARENT_ID,
					WORK_PACKAGE_JOURNALS.RESPONSIBLE_ID,
					WORK_PACKAGE_JOURNALS.BUDGET_ID,
					WORK_PACKAGE_JOURNALS.STORY_POINTS,
					WORK_PACKAGE_JOURNALS.REMAINING_HOURS,
					WORK_PACKAGE_JOURNALS.DERIVED_ESTIMATED_HOURS,
					WORK_PACKAGE_JOURNALS.SCHEDULE_MANUALLY,
					WORK_PACKAGE_JOURNALS.DURATION,
					WORK_PACKAGE_JOURNALS.IGNORE_NON_WORKING_DAYS,
					WORK_PACKAGE_JOURNALS.DERIVED_REMAINING_HOURS,
					WORK_PACKAGE_JOURNALS.DERIVED_DONE_RATIO,
					WORK_PACKAGE_JOURNALS.PROJECT_PHASE_DEFINITION_ID
				FROM
					JOURNALS
					INNER JOIN WORK_PACKAGE_JOURNALS ON "journals"."data_type" = 'Journal::WorkPackageJournal'
					AND "journals"."data_id" = "work_package_journals"."id"
				AND (
				VALIDITY_PERIOD @> TIMESTAMP WITH TIME ZONE '2025-09-10 12:00:00 UTC'
				OR VALIDITY_PERIOD @> TIMESTAMP WITH TIME ZONE '2025-10-10 11:32:01 UTC'
			    )
			),
			... the rest remains unchanged
```

Doing so solves the problem while at the same time greatly simplifying the code because the substitution happens at one central place within the SQL.

The only place which still needs to be adapted are WHERE statements on custom values which still need to be rewritten.

To ensure that there are no other failures because of the alternative `WorkPackage.allowed_to` path, the whole test suite was run with the alternative branch active (using `large_instance_wp_allowed_to_sql: true`). It passed. That change has been reverted. Ideally, once and if it has been established that the alternative branch is to be preferred the alternative branch should become the only one used. If not, the alternative branch should be deleted. Doing so will prevent regressions.

# Merge checklist

- [x] Added/updated tests

ulferts committed:
spec baseline with and without allowed_to perf improvement

ulferts committed:
TO BE REVOKED - enable allowed_to perf improvement in all specs

ulferts committed:
rewrite allowed_to to avoid from

ulferts committed:
provide a cte shadowing the work_packages table for historic query

ulferts committed:
cleanup now unused code

ulferts committed:
document the approach

ulferts committed:
simplify super call

ulferts committed:
Revert "TO BE REVOKED - enable allowed_to perf improvement in all specs"

This reverts commit 0c104aee0c73ba8bd7847701f3b0db19d000eb95.

toy commented on app/models/journable/historic_active_record_relation.rb:
```diff
@@ -65,145 +58,108 @@ def initialize(relation, timestamp:)
+      if model.column_names_missing_in_journal.include?(column_name.to_s)
+        Rails.logger.warn "Cannot pluck column `#{column_name}` because this attribute is not journalized," \
+                          "i.e. it is missing in the #{journal_class.table_name} table."
+        "null as #{column_name}"
```
Maybe better to explicitly wrap with `Arel.sql`?
```suggestion
        Arel.sql("null as #{column_name}")
```

toy commented on app/models/work_packages/scopes/allowed_to.rb:
Can use `SELECT id` instead of `SELECT *`, probably not for optimisation, but for clarity:
```suggestion
          SELECT id from work_packages
```

toy commented on spec/models/journable/historic_active_record_relation_spec.rb:
```diff
@@ -124,12 +124,6 @@
 
-      describe "#to_sql" do
-        it "transforms the expression to query the correct table" do
-          expect(subject.to_sql).to include "\"work_package_journals\".\"project_id\" IN (#{project.id}, 1, 2, 3)"
```
This and following can check that CTE redefining `work_packages` is used

ulferts replied:
In the past, I did implement specs checking the resulting SQL. And while I think they have value when developing the feature, their value declines after the feature is done. While one is developing, you are very deep in the matter and have a goal in mind which is to get the SQL to work because you know that this will result in the expected actual results. But it is always just an intermediary because you actually want to check that something else is working. So I'd rather check for that, which I think the tests already do by expecting certain values, than checking for the SQL. On top of that, the SQL might change if a library decides to switch e.g. the order of statements or whitespaces. And those changes are irrelevant as long as the actual result is achieved.

toy replied:
I see those deleted tests as a way to check that in all cases the SQL to query data from journals was used. Though, it is better to check that we receive models from journals, which is why I suggested in other comment for this file to check for `journal_id`. For `pluck` tests there is an explicit check in most cases for attributes that are different from the value in the model, but for `where` and others there is only a check for presence of model in the output (equality is checked by class and id), so I think it would be better to ensure that an "enhanced" model from journal is returned

ulferts replied:
Ok, I can get behind that idea. 

I therefore added specs to check for the enhanced model being returned.

toy commented on spec/models/journable/historic_active_record_relation_spec.rb:
```diff
@@ -124,12 +124,6 @@
-        end
-      end
-
       describe "#to_a" do
```
Is it maybe better to check that resulting model is not normal WorkPackage, but contains `journal_id` or something like that?

toy commented on spec/models/journable/timestamps_spec.rb:
```diff
@@ -273,20 +273,6 @@
         describe "when querying with at_timestamp" do
           subject { relation.at_timestamp(friday) }
 
-          it "joins the journable table rather than the journal-data table" do
```
Also can check for CTE usage to redefine `work_packages`

toy commented on spec/requests/api/v3/work_packages/index_resource_spec.rb:
```diff
@@ -244,7 +244,7 @@
     end
 
-    context "when providing timestamps", with_ee: %i[baseline_comparison] do
+    shared_examples_for "when providing timestamps" do
```
As it is required for examples, `with_ee: %i[baseline_comparison]` can also stay on shared_example, then also outer `describe "baseline comparison"` is not necessary

ulferts replied:
Thanks, good tip

toy commented on app/models/work_packages/scopes/allowed_to.rb:
You mentioned it in the description, but if I understand correctly, this still will make `project_id` be the one from journal instead of from `work_packages` table, or am I missing something?

ulferts replied:
Not in the normal application of the scope. But when applied to an SQL generated from HistoricActiveRecordRelation, since `work_packages` will in fact be `work_package_journals`, then yes. There is a spec that enforces that.

toy replied:
So it is expected and intended

toy commented on app/models/journable/historic_active_record_relation.rb:
```diff
@@ -499,49 +217,11 @@ def timestamp_case_when_statements
                           timestamp.in_time_zone
                         else
-                          raise NotImplementedError, "Unknown timestamp type: #{timestamp.class}"
+                          raise NoMethodError, "Unknown timestamp type: #{timestamp.class}"
```
👍

toy commented on app/models/journable/historic_active_record_relation.rb:
```diff
@@ -499,49 +217,11 @@ def timestamp_case_when_statements
-    "#{statement} AND (#{additional_conditions.join(' OR ')})"
-  end
-
-  class NotImplementedError < StandardError; end
```
This was a very confusing definition

ulferts replied:
TBH, I think it was done to circumvent the shortcomings of NotImplementedError.

toy commented on app/models/journable/historic_active_record_relation.rb:
Is it possible to also redefine `custom_values` using CTE to not need at least part of this code?

ulferts replied:
I tried that but didn't manage so far. The limiting factor is that there might be multiple entries/states returned for a single work package. And the joining currently happens on the id of the work package. Before further optimization s are done, the way the custom fields are filtered for (without the historic aspect) should be reworked, IMO.

toy commented on app/models/journable/historic_active_record_relation.rb:
```diff
@@ -65,145 +58,108 @@ def initialize(relation, timestamp:)
-    relation = add_join_projects_on_work_package_journals(relation)
-    relation = select_columns_from_the_appropriate_tables(relation)
+  def build_arel(connection, aliases = nil)
+    substitute_join_tables_in_where_clause(self)
```
Not sure if it is safe in all cases to modify predicates inplace, probably better to at least add bang to the method name

ulferts replied:
I am also not sure about this 😆 . But the method is overwritten, so I need to keep `build_arel`. Or do you mean `substitute_join_tables_in_where_clause`? Given that it is an internal method, I don't think this needs to be communicated like that.

toy replied:
I meant for `substitute_join_tables_in_where_clause`

toy commented on app/models/journable/historic_active_record_relation.rb:
Return value is unused

ulferts replied:
👍

toy commented on app/models/journable/historic_active_record_relation.rb:
```diff
@@ -65,145 +58,108 @@ def initialize(relation, timestamp:)
 
-  private
+  def historic_models_statement
+    relation = Journal
+                 .joins(historic_models_data_journals_join)
+                 .joins(historic_models_original_models_join)
+                 .select(historic_models_selects)
```
It is possible to at least partially join using dsl (`model.joins(:journals)`, `model.journal_class.joins(:journal)`), but I failed to find a way to combine those

toy commented on app/models/journable/historic_active_record_relation.rb:
```diff
@@ -65,145 +58,108 @@ def initialize(relation, timestamp:)
+    <<-SQL.squish
+      INNER JOIN #{model.journal_class.table_name}
+      ON "journals"."data_type" = '#{model.journal_class.name}'
+      AND "journals"."data_id" = "#{model.journal_class.table_name}"."id"
```
In some cases `#{Journal.table_name}` is used, in others `"jounrals"` explicitly

toy commented on app/models/journable/historic_active_record_relation.rb:
```diff
@@ -65,145 +58,108 @@ def initialize(relation, timestamp:)
+  end
+
+  def historic_models_selects
+    ["journals.journable_id AS id",
```
Also explicit `"journals"`

ulferts committed:
always use table_name

ulferts committed:
wrap literal sql in Arel

ulferts committed:
remove unused return value

ulferts committed:
limit projection to used column

ulferts committed:
move with_ee statement

ulferts commented:
@toy , thanks for the review. I think I addressed all the points raised by you except for the one about testing the SQL which I don't place value in. I wrote an explanation on the why next to your comment.

Can you please recheck?

toy commented on app/models/journable/historic_active_record_relation.rb:
```diff
@@ -65,145 +58,110 @@
+      if model.column_names_missing_in_journal.include?(column_name.to_s)
+        Rails.logger.warn "Cannot pluck column `#{column_name}` because this attribute is not journalized," \
+                          "i.e. it is missing in the #{journal_class.table_name} table."
+        Arel.sql("null as #{column_name}")
```
Will it be happier about `Arel::Nodes::SqlLiteral.new('NULL').as(column_name)`?

toy commented on app/models/journable/historic_active_record_relation.rb:
```diff
@@ -223,273 +180,38 @@ def substitute_join_tables_in_where_clause(relation)
-      # The customizable_journals table has no direct relation to the work_packages table,
-      # but it has to the journals table. We join it to the journals table instead.
-      journal_id = "customizable_journals.journal_id = journals.id"
+    predicate.gsub! /JOIN (?<!_)#{custom_values}/,
```
Why is negative look behind for `_` needed?

toy commented on app/models/journable/historic_active_record_relation.rb:
```diff
@@ -223,273 +180,38 @@
-      # The customizable_journals table has no direct relation to the work_packages table,
-      # but it has to the journals table. We join it to the journals table instead.
-      journal_id = "customizable_journals.journal_id = journals.id"
+    predicate.gsub! /JOIN (?<!_)#{custom_values}/,
```
Maybe `Regexp.escape` on table names would be appropriate?

ulferts replied:
This is a false positive. `custom_values` is static and will always be `custom_values = CustomValue.table_name`. I therefore see no risk here.

toy reviewed (approved):
Looks good, see comments, but I didn't notice anything major

ulferts committed:
attempt to sooth SQL code injection warning

ulferts committed:
bolster specs so that they ensure working with historic work packages
